File: CHTL/CHTLParser/Parser.cpp
#include "Parser.hpp"
#include <iostream>

namespace CHTL {

Parser::Parser(Lexer& l) : lexer(l) {
    nextToken();
    nextToken();
}

void Parser::nextToken() {
    currentToken = peekToken;
    peekToken = lexer.nextToken();
}

bool Parser::expectPeek(TokenType t) {
    if (peekToken.type == t) {
        nextToken();
        return true;
    }
    return false;
}

const std::vector<std::string>& Parser::getErrors() const {
    return errors;
}

std::unique_ptr<Program> Parser::parseProgram() {
    auto program = std::make_unique<Program>();
    while (currentToken.type != TokenType::END_OF_FILE) {
        auto node = parseNode();
        if (node) {
            program->addChild(std::move(node));
        } else {
            nextToken();
        }
    }
    return program;
}

std::unique_ptr<Node> Parser::parseNode() {
    if (currentToken.type == TokenType::COMMENT) {
        auto node = parseCommentNode();
        return node;
    }
    if (currentToken.type == TokenType::IDENT) {
        auto node = parseElementNode();
        return node;
    }
    return nullptr;
}

void Parser::parseAttributes(ElementNode* element) {
    std::string key = currentToken.literal;

    if (!expectPeek(TokenType::COLON) && !expectPeek(TokenType::EQUAL)) {
        errors.push_back("Expected ':' or '=' after attribute name.");
        return;
    }

    nextToken();

    std::string value;
    if (currentToken.type == TokenType::STRING || currentToken.type == TokenType::IDENT) {
        value = currentToken.literal;
    } else {
        errors.push_back("Attribute value must be a string or an identifier.");
        return;
    }

    element->attributes[key] = value;

    if (peekToken.type == TokenType::SEMICOLON) {
        nextToken();
    }
}

void Parser::parseStyleBlock(ElementNode* parent) {
    nextToken(); // consume 'style'
    if (currentToken.type != TokenType::LBRACE) {
        errors.push_back("Expected '{' after 'style' identifier.");
        return;
    }
    nextToken();

    while (currentToken.type != TokenType::RBRACE && currentToken.type != TokenType::END_OF_FILE) {
        if (currentToken.type != TokenType::IDENT) {
            errors.push_back("Expected style property name.");
            break;
        }
        std::string key = currentToken.literal;

        if (!expectPeek(TokenType::COLON)) {
            errors.push_back("Expected ':' after style property name.");
            break;
        }
        nextToken();

        std::string value;
        while (currentToken.type != TokenType::SEMICOLON && currentToken.type != TokenType::RBRACE) {
            value += currentToken.literal;
            nextToken();
        }

        parent->styles[key] = value;

        if (currentToken.type == TokenType::SEMICOLON) {
            nextToken();
        }
    }
}

std::unique_ptr<ElementNode> Parser::parseElementNode() {
    auto elementNode = std::make_unique<ElementNode>(currentToken.literal);
    nextToken();

    if (elementNode->tagName == "text") {
        if (currentToken.type != TokenType::LBRACE) {
            errors.push_back("Expected '{' after 'text' identifier.");
            return nullptr;
        }
        nextToken();

        std::string textContent;
        if (currentToken.type == TokenType::STRING) {
            textContent = currentToken.literal;
            nextToken();
        } else {
            while(currentToken.type != TokenType::RBRACE && currentToken.type != TokenType::END_OF_FILE) {
                textContent += currentToken.literal;
                nextToken();
                if (currentToken.type != TokenType::RBRACE) {
                    textContent += " ";
                }
            }
        }
        auto textNode = std::make_unique<TextNode>(textContent);
        elementNode->addChild(std::move(textNode));

        if (currentToken.type != TokenType::RBRACE) {
            errors.push_back("Expected '}' to close text block.");
            return nullptr;
        }
        return elementNode;
    }

    if (currentToken.type != TokenType::LBRACE) {
        errors.push_back("Expected '{' after element identifier.");
        return nullptr;
    }
    nextToken();

    while (currentToken.type != TokenType::RBRACE && currentToken.type != TokenType::END_OF_FILE) {
        if (currentToken.type == TokenType::COMMENT) {
            auto childNode = parseCommentNode();
            if (childNode) elementNode->addChild(std::move(childNode));
        } else if (currentToken.type == TokenType::IDENT) {
            if (currentToken.literal == "style") {
                parseStyleBlock(elementNode.get());
            } else if (peekToken.type == TokenType::COLON || peekToken.type == TokenType::EQUAL) {
                parseAttributes(elementNode.get());
            } else {
                auto childNode = parseNode();
                if (childNode) elementNode->addChild(std::move(childNode));
                continue;
            }
        } else {
            errors.push_back("Unexpected token inside element block: " + currentToken.literal);
        }
        nextToken();
    }

    return elementNode;
}

std::unique_ptr<CommentNode> Parser::parseCommentNode() {
    return std::make_unique<CommentNode>(currentToken.literal);
}

} // namespace CHTL

File: CHTL/CHTLParser/parser.py
from CHTL.CHTLLexer.token import Token
from CHTL.CHTLLexer.token_type import TokenType
from CHTL.CHTLNode import nodes

class ParseError(Exception): pass

class Parser:
    def __init__(self, tokens: list[Token]):
        self.tokens = tokens
        self.current = 0

    def parse(self) -> nodes.ProgramNode:
        program = nodes.ProgramNode()
        while not self._is_at_end():
            node = self._declaration()
            if node: program.children.append(node)
        return program

    def _declaration(self):
        if self._peek().type == TokenType.IDENTIFIER:
            return self._element_statement()
        if self._peek().type == TokenType.EOF:
            return None
        # Basic error recovery
        self._advance()
        return None

    def _element_statement(self):
        tag_name = self._consume(TokenType.IDENTIFIER, "element tag name").lexeme
        element = nodes.ElementNode(tag_name=tag_name)

        if not self._match(TokenType.LBRACE):
            self._consume(TokenType.SEMICOLON, "Expect '{' or ';' after element name.")
            return element

        while not self._check(TokenType.RBRACE) and not self._is_at_end():
            if self._check(TokenType.TEXT):
                element.children.append(self._text_statement())
            elif self._is_attribute():
                element.attributes.append(self._attribute())
            elif self._check(TokenType.IDENTIFIER):
                element.children.append(self._element_statement())
            else:
                self._error(self._peek(), "Unexpected token inside element block.")
                self._advance() # Skip unexpected token

        self._consume(TokenType.RBRACE, "}")
        return element

    def _text_statement(self):
        self._consume(TokenType.TEXT, "'text'")
        self._consume(TokenType.LBRACE, "{")
        token = self._consume(TokenType.STRING, "string literal for text content")
        value = token.lexeme
        self._consume(TokenType.RBRACE, "}")
        return nodes.TextNode(value=value)

    def _attribute(self):
        name = self._consume(TokenType.IDENTIFIER, "attribute name").lexeme
        if not self._match(TokenType.EQUALS, TokenType.COLON):
            raise self._error(self._peek(), "Expect '=' or ':' for attribute assignment.")

        if self._match(TokenType.STRING, TokenType.IDENTIFIER):
            value = self._previous().lexeme
        else:
            raise self._error(self._peek(), "Expect string or identifier for attribute value.")

        self._consume(TokenType.SEMICOLON, "';'")
        return nodes.AttributeNode(name=name, value=value)

    def _is_attribute(self):
        return self._check(TokenType.IDENTIFIER) and self._peek(1).type in [TokenType.COLON, TokenType.EQUALS]

    def _consume(self, ttype: TokenType, message: str) -> Token:
        if self._check(ttype): return self._advance()
        raise self._error(self._peek(), f"Expect {message}")

    def _check(self, ttype: TokenType) -> bool:
        if self._is_at_end(): return False
        return self._peek().type == ttype

    def _is_at_end(self, offset=0) -> bool:
        return self.tokens[self.current + offset].type == TokenType.EOF

    def _peek(self, offset=0) -> Token:
        return self.tokens[self.current + offset]

    def _previous(self) -> Token:
        return self.tokens[self.current - 1]

    def _advance(self) -> Token:
        if not self._is_at_end(): self.current += 1
        return self._previous()

    def _match(self, *types: TokenType) -> bool:
        for ttype in types:
            if self._check(ttype):
                self._advance()
                return True
        return False

    def _error(self, token: Token, message: str) -> ParseError:
        # For now, just print the error and allow parsing to continue
        print(f"[Line {token.line}] Error at '{token.lexeme}': {message}")
        return ParseError()
